apps/emqx_durable_storage/src/emqx_ds_lts.erl

```diff
@@ -0,0 +1,509 @@
+    {1, S111} = trie_insert(T, S11, <<"bar">>),
+    ?assertMatch({false, S111}, trie_next(T, S11, <<"bar">>)).
+
+lookup_key_test() ->
+    T = trie_create(),
+    {_, S1} = trie_insert(T, ?PREFIX, <<"foo">>),
+    {_, S11} = trie_insert(T, S1, <<"foo">>),
+    %% Topics don't match until we insert ?EOT:
+    ?assertMatch( undefined
+                , lookup_topic_key(T, [<<"foo">>])
+                ),
+    ?assertMatch( undefined
+                , lookup_topic_key(T, [<<"foo">>, <<"foo">>])
+                ),
+    {_, S10} = trie_insert(T, S1, ?EOT),
+    {_, S110} = trie_insert(T, S11, ?EOT),
+    ?assertMatch( {ok, {S10, []}}
+                , lookup_topic_key(T, [<<"foo">>])
+                ),
+    ?assertMatch( {ok, {S110, []}}
+                , lookup_topic_key(T, [<<"foo">>, <<"foo">>])
+                ),
+    %% The rest of keys still don't match:
+    ?assertMatch( undefined
+                , lookup_topic_key(T, [<<"bar">>])
+                ),
+    ?assertMatch( undefined
+                , lookup_topic_key(T, [<<"bar">>, <<"foo">>])
+                ).
+
+wildcard_lookup_test() ->
+    T = trie_create(),
+    {1, S1} = trie_insert(T, ?PREFIX, <<"foo">>),
+    {0, S11} = trie_insert(T, S1, ?PLUS), %% Plus doesn't increase the number of children
+    {1, S111} = trie_insert(T, S11, <<"foo">>),
+    {0, S1110} = trie_insert(T, S111, ?EOT), %% ?EOT doesn't increase the number of children
+    ?assertMatch( {ok, {S1110, [<<"bar">>]}}
+                , lookup_topic_key(T, [<<"foo">>, <<"bar">>, <<"foo">>])
+                ),
+    ?assertMatch( {ok, {S1110, [<<"quux">>]}}
+                , lookup_topic_key(T, [<<"foo">>, <<"quux">>, <<"foo">>])
+                ),
+    ?assertMatch( undefined
+                , lookup_topic_key(T, [<<"foo">>])
+                ),
+    ?assertMatch( undefined
+                , lookup_topic_key(T, [<<"foo">>, <<"bar">>])
+                ),
+    ?assertMatch( undefined
+                , lookup_topic_key(T, [<<"foo">>, <<"bar">>, <<"bar">>])
+                ),
+    ?assertMatch( undefined
+                , lookup_topic_key(T, [<<"bar">>, <<"foo">>, <<"foo">>])
+                ),
+    {_, S10} = trie_insert(T, S1, ?EOT),
+    ?assertMatch( {ok, {S10, []}}
+                , lookup_topic_key(T, [<<"foo">>])
+                ).
+
+%% erlfmt-ignore
+topic_key_test() ->
+    T = trie_create(),
+    try
+        Threshold = 3,
+        ThresholdFun = fun(0) -> 1000;
+                          (_) -> Threshold
+                       end,
+        %% Test that bottom layer threshold is high:
+        lists:foreach(
+          fun(I) ->
+                  {_, []} = test_key(T, ThresholdFun, [I, 99, 99, 99])
+          end,
+          lists:seq(1, 10)),
+        %% Test adding children on the 2nd level:
+        lists:foreach(
+          fun(I) ->
+                  case test_key(T, ThresholdFun, [1, I, 1]) of
+                      {_, []} when I < Threshold ->
```

Comment:
Q: perhaps it's this one that has the [one-off error](https://github.com/emqx/emqx/pull/11712/files#diff-6964e4d7d9b684f01e4fc035e0b5d23617e718f75de5ec2b0125cd0514540e7dR410), because a previous assertion (`[1,99,...])` inserted a key that counted towards the threshold?

It _seems_ from the definition of `do_topic_key` that the `?PLUS` level is inserted only _after_ the threshold is crossed, so we would expect this line to be `when I =< Threshold` if `[1,99, ...]` hadn't been inserted?